File: modules/catalog.js
import { globbyStream } from "globby"
import { readFile } from "fs/promises"
import { load } from "js-yaml"
import chalk from "chalk"
import { $ } from "zx"

const meta = {
  // Piece subtitle (sub)
  ukrfs: "Українська народна пісня",
  ukrfsvar: "Варіації на тему української народної пісні",
  ukrfd: "Український народний танець",
  rusfs: "Російська народна пісня",
  rusfsvar: "Варіації на тему російської народної пісні",
  rusfd: "Російський народний танець",
  hunfs: "Угорська народна пісня",
  hunfd: "Угорський народний танець",
  mdafs: "Молдавська народна пісня",
  mdafd: "Молдавський народний танець",
  roufs: "Румуньска народна пісня",
  roufd: "Румунський народний танець",
  blrfs: "Білоруська народна пісня",
  blrfd: "Білоруський народний танець",
  polfs: "Польська народна пісня",
  polfd: "Польський народний танець",
  czefs: "Чеська народна пісня",
  czefd: "Чеський народний танець",
  // Piece arrangment type (art)
  arr: "Обр.", // обробка = arrangment (default)
  ipr: "Пер.", // переклад = interpratation
  hrm: "Гарм." // гармонізація = harmonization
}

function addScoreFile(score) {
  const title = score.tit.replace(/[',!?]/g, "").replace(/ /g, "-")
  score.file = `${title}-${score.id}`
}

function addPieceMeta(piece) {
  if (piece.sub in meta) { piece.sub = meta[piece.sub] }
  if (piece.arr) {
    piece.art = piece.art ? meta[piece.art] : meta.arr
  }
}

function addSectionPieces(section, cpieces) {
  const spieces = []
  for (const spiece of section.pieces) {
    if (cpieces.has(spiece)) {
      spieces.push(cpieces.get(spiece))
    } else {
      throw new Error(`Piece ${spiece} is not in catalog`)
    }
  }
  section.pieces = spieces
}

function addBookPieces(book, cpieces) {
  const bpieces = []
  for (const bpiece of book.pieces) {
    if (bpiece.sec) {
      addSectionPieces(bpiece, cpieces)
      bpieces.push(bpiece)
    } else if (cpieces.has(bpiece)) {
      bpieces.push(cpieces.get(bpiece))
    } else {
      throw new Error(`Piece ${bpiece} is not in catalog`)
    }
  }
  book.pieces = bpieces
}

function countBookPieces(book) {
  let pieceCount = 0
  for (const piece of book.pieces) {
    if (piece.sec) { pieceCount += piece.pieces.length }
    else { ++pieceCount }
  }
  return pieceCount
}

export async function readPieces(args, all = false) {
  const pieces = []
  let pieceCount = 0
  const glob = `catalog/*${args.c}*.yaml`
  for await (const catalog of globbyStream(glob)) {
    const { pieces: cpieces } = load(await readFile(catalog))
    for (const piece of cpieces) {
      ++pieceCount
      if (all || args._.includes("all") || args._.includes(piece.id)) {
        addScoreFile(piece)
        addPieceMeta(piece)
        pieces.push(piece)
      }
    }
  }
  console.log(
    chalk.blue("Catalog:"),
    chalk.yellow.bold(`${pieceCount} pieces`)
  )
  return pieces
}

async function readPiecesFromBook(args) {
  function addBookPiece(bpiece) {
    if (cpieces.has(bpiece)) {
      bpieces.push(cpieces.get(bpiece))
    } else {
      throw new Error(`Piece ${bpiece} is not in catalog`)
    }
  }
  const pieces = await readPieces(args, true)
  const cpieces = new Map()
  pieces.forEach(piece => cpieces.set(piece.id, piece))
  let { books } = load(await readFile("meta/books.yaml"))
  const bpieces = []
  for (const book of books) {
    if (args._.includes("all") || args._.includes(book.id)) {
      for (const bpiece of book.pieces) {
        if (bpiece.sec) {
          for (const spiece of bpiece.pieces) { addBookPiece(spiece) }
        } else { addBookPiece(bpiece) }
      }
    }
  }
  return bpieces
}

export async function readBooks(args) {
  const pieces = await readPieces(args, true)
  const cpieces = new Map()
  pieces.forEach(piece => cpieces.set(piece.id, piece))
  const books = []
  let { books: cbooks } = load(await readFile("meta/books.yaml"))
  for (const book of cbooks) {
    if (args._.includes("all") || args._.includes(book.id)) {
      addScoreFile(book)
      addBookPieces(book, cpieces)
      books.push(book)
      console.log(
        chalk.blue(`Book ${book.id}:`),
        chalk.yellow.bold(`${countBookPieces(book)} pieces`)
      )
    }
  }
  return books
}

export async function selectPieces(args) {
  const match = (tag) => (piece) => {
    if (!args[tag]) { return true }
    if (!piece[tag]) { return false }
    const matchers = args[tag].split(",").map(query => label => {
      const neg = query.startsWith("^")
      const re = new RegExp(neg ? query.slice(1) : query, "i")
      return neg ? !re.test(label) : re.test(label)
    })
    const labels = [].concat(piece[tag])
    return matchers.some(match => labels.some(match))
  }
  const tags = [
    "org", "sty", "gnr", "ton", "frm", "bss", "lvl", "tit", "com", "arr"
  ]
  const matchers = tags.map(match)
  const pieces = args.frbk ?
        await readPiecesFromBook(args) :
        await readPieces(args)
  const selected = pieces.filter(piece =>
    matchers.every(match => match(piece))
  )
  console.log(
    chalk.blue("Selected:"),
    chalk.yellow.bold(`${selected.length} pieces`)
  )
  return selected
}

export async function playPieces(pieces, args) {
  function* playIndex() {
    let i = 0, len = pieces.length, played = new Set()
    while (args.cycle ? true : i < len) {
      if (args.random) {
        if (played.size === len) { played.clear() }
        let ri = Math.floor(Math.random() * len)
        while (played.has(ri)) { ri = Math.floor(Math.random() * len) }
        played.add(ri)
        yield (++i, ri)
      } else { yield i++ % len }
    }
  }
  for (const i of playIndex()) {
    const { id, tit, file, com, arr, art, lvl } = pieces[i]
    if (args.dry) {
      const summary = [chalk.green(id), chalk.yellow(tit)]
      if (com) { summary.push(chalk.cyan(com)) }
      if (arr) { summary.push(chalk.magenta(art), chalk.magenta(arr)) }
      summary.push(lvl)
      console.log(summary.join(" "))
      continue
    }
    await $`xdg-open scores/${file}.pdf`
  }
}

File: internal/catalog/catalog.go
package catalog

import (
  "fmt"
  "strings"
  "regexp"
  "path/filepath"
  "os"
  "github.com/sanity-io/litter"
  "gopkg.in/yaml.v3"
)

var meta = map[string]string{
  // Piece subtitle (sub)
  "ukrfs": "Українська народна пісня",
  "ukrfsvar": "Варіації на тему української народної пісні",
  "ukrfd": "Український народний танець",
  "rusfs": "Російська народна пісня",
  "rusfsvar": "Варіації на тему російської народної пісні",
  "rusfd": "Російський народний танець",
  "blrfs": "Білоруська народна пісня",
  "blrfd": "Білоруський народний танець",
  "hunfs": "Угорська народна пісня",
  "hunfd": "Угорський народний танець",
  "mdafs": "Молдавська народна пісня",
  "mdafd": "Молдавський народний танець",
  "roufs": "Румуньска народна пісня",
  "roufd": "Румунський народний танець",
  "polfs": "Польська народна пісня",
  "polfd": "Польський народний танець",
  "czefs": "Чеська народна пісня",
  "czefd": "Чеський народний танець",
  "svkfs": "Словацька народна пісня",
  "svkfd": "Словацький народний танець",
  // Piece arrangment type (art)
  "arr": "Обр.", // обробка = arrangment (default)
  "ipr": "Пер.", // переклад = interpratation
  "hrm": "Гарм.", // гармонізація = harmonization
}

type StrSlice []string

func (ss *StrSlice) UnmarshalYAML(node *yaml.Node) error {
  var slc []string
  err := node.Decode(&slc) // try a string slice
  if err != nil {
    var str string
    err := node.Decode(&str) // try a single string
    if err != nil {
      return err
    }
    *ss = []string{str} // convert a single string to a string slice
    return nil
  }
  *ss = slc // return a string slice
  return nil
}

type Piece struct {
  ID string `yaml:"id"`
  Tit string `yaml:"tit"`
  Sub string `yaml:"sub"`
  Com string `yaml:"com"`
  Arr string `yaml:"arr"`
  Art string `yaml:"art"`
  Src string `yaml:"src"`
  Org string `yaml:"org"`
  Sty string `yaml:"sty"`
  Gnr string `yaml:"gnr"`
  Ton StrSlice `yaml:"ton"`
  Frm StrSlice `yaml:"frm"`
  Bss StrSlice `yaml:"bss"`
  Lvl string `yaml:"lvl"`
  File string
}

type PieceMap map[string]Piece

type RawSection struct {
  Tit string `yaml:"tit"`
  Pieces []string `yaml:"pieces"`
}

type RawBook struct {
  ID string `yaml:"id"`
  Tit string `yaml:"tit"`
  Sub string `yaml:"sub"`
  Pieces []string `yaml:"pieces"`
  Sections []RawSection `yaml:"sections"`
}

type Section struct {
  Tit string
  Pieces []Piece
}

type Book struct {
  ID string
  Tit string
  Sub string
  Pieces []Piece
  Sections []Section
}

type PieceQueries map[string]string

type PlayCommand struct {
  Catalog string
  Book, Cycle, Random, List, All bool
  Pieces, Books []string
  Queries PieceQueries
}

type MatchStr func(str string) bool

func makeMatchStr(pattern string) (MatchStr, error) {
  isNeg := strings.HasPrefix(pattern, "^")
  pattern = regexp.QuoteMeta(strings.TrimPrefix(pattern, "^"))
  re, err := regexp.Compile(strings.ReplaceAll(pattern, ",", "|"))
  if err != nil {
    return nil, err
  }
  match := func (str string) bool {
    if isNeg {
      return !re.MatchString(str)
    } else {
      return re.MatchString(str)
    }
  }
  return match, nil
}

type MatchPiece func(piece Piece) bool

func makeMatchPiece(queries PieceQueries) (MatchPiece, error) {
  matches := make(map[string]MatchStr, len(queries))
  for opt, query := range queries {
    match, err := makeMatchStr(query)
    if err != nil {
      return nil, err
    }
    matches[opt] = match
  }
  matchSlice := func(slc StrSlice, match MatchStr) bool {
    for _, str := range slc {
      if match(str) {
        return true
      }
    }
    return false
  }
  matchPiece := func(piece Piece) bool {
    for opt, match := range matches {
      switch opt {
      case "org": if !match(piece.Org) { return false }
      case "sty": if !match(piece.Sty) { return false }
      case "gnr": if !match(piece.Gnr) { return false }
      case "ton": if !matchSlice(piece.Ton, match) { return false }
      case "frm": if !matchSlice(piece.Frm, match) { return false }
      case "bss": if !matchSlice(piece.Bss, match) { return false }
      case "lvl": if !match(piece.Lvl) { return false }
      case "tit": if !match(piece.Tit) { return false }
      case "com": if !match(piece.Com) { return false }
      case "arr": if !match(piece.Arr) { return false }
      default: panic(fmt.Sprintf("catalog: unknown query option %v", opt))
      }
    }
    return true
  }
  return matchPiece, nil
}

func listCatalogFiles(catDir, catQuery string) ([]string, error) {
  entries, err := os.ReadDir(catDir)
  if err != nil {
    return nil, err
  }
  match, err := makeMatchStr(catQuery)
  if err != nil {
    return nil, err
  }
  files := make([]string, 0, 50)
  for _, entry := range entries {
    if !entry.IsDir() && strings.HasSuffix(entry.Name(), ".yaml") &&
      match(strings.TrimSuffix(entry.Name(), ".yaml")) {
      files = append(files, entry.Name())
    }
  }
  return files, nil
}

func readCatalogFile(catDir, catFile string) ([]Piece, error) {
  file, err := os.Open(filepath.Join(catDir, catFile))
  if err != nil {
    return nil, err
  }
  defer file.Close()
  dec := yaml.NewDecoder(file)
  var pieces struct { Pieces []Piece `yaml:"pieces"` }
  err = dec.Decode(&pieces)
  if err != nil {
    return nil, err
  }
  return pieces.Pieces, nil
}

func addMetaToPieces(pieces []Piece) {
  cleanTit := regexp.MustCompile(`[',\.!\?]`)
  for i := range pieces {
    piece := &pieces[i]
    // sub
    if sub, ok := meta[piece.Sub]; ok {
      piece.Sub = sub
    }
    // art
    if len(piece.Arr) > 0 {
      if len(piece.Art) > 0 {
        piece.Art = meta[piece.Art]
      } else {
        piece.Art = meta["arr"] // default: arrangement
      }
    }
    // file
    tit := strings.ReplaceAll(
      cleanTit.ReplaceAllLiteralString(piece.Tit, ""), " ", "-",
    )
    piece.File = fmt.Sprintf("%v-%v", tit, piece.ID)
  }
}

func readPieces(catDir, catQuery string) (PieceMap, error) {
  catFiles, err := listCatalogFiles(catDir, catQuery)
  if err != nil {
    return nil, err
  }
  pieceMap := make(PieceMap, 1000)
  for _, catFile := range catFiles {
    pieces, err := readCatalogFile(catDir, catFile)
    if err != nil {
      return nil, err
    }
    addMetaToPieces(pieces)
    for _, piece := range pieces {
      pieceMap[piece.ID] = piece
    }
  }
  return pieceMap, nil
}

func readBookFile(bookDir, bookFile string) ([]RawBook, error) {
  file, err := os.Open(filepath.Join(bookDir, bookFile))
  if err != nil {
    return nil, err
  }
  defer file.Close()
  dec := yaml.NewDecoder(file)
  var books struct { Books []RawBook `yaml:"books"` }
  err = dec.Decode(&books)
  if err != nil {
    return nil, err
  }
  return books.Books, nil
}

func selectRawBooks(rawBooks []RawBook, bookIDs []string) ([]RawBook, error) {
  rawBookMap := make(map[string]RawBook, 50)
  for _, rawBook := range rawBooks {
    rawBookMap[rawBook.ID] = rawBook
  }
  selRawBooks := make([]RawBook, 0, 50)
  for _, bookID := range bookIDs {
    if rawBook, ok := rawBookMap[bookID]; ok {
      selRawBooks = append(selRawBooks, rawBook)
    } else {
      return nil, fmt.Errorf("invalid book %v", bookID)
    }
  }
  return selRawBooks, nil
}

func addPiecesToBooks(rawBooks []RawBook, pieceMap PieceMap) ([]Book, error) {
  books := make([]Book, 0, 50)
  for _, rawBook := range rawBooks {
    book := Book{ID: rawBook.ID, Tit: rawBook.Tit, Sub: rawBook.Sub}
    book.Pieces = make([]Piece, 0, 200)
    if len(rawBook.Sections) > 0 { // book with sections
      book.Sections = make([]Section, 0, 10)
      for _, rawSec := range rawBook.Sections {
        sec := Section{Tit: rawSec.Tit}
        sec.Pieces = make([]Piece, 0, 20)
        for _, pieceID := range rawSec.Pieces {
          if piece, ok := pieceMap[pieceID]; ok {
            sec.Pieces = append(sec.Pieces, piece)
            book.Pieces = append(book.Pieces, piece)
          } else {
            err := fmt.Errorf(
              "piece %v from book %v is not in catalog", pieceID, book.ID,
            )
            return nil, err
          }
        }
        book.Sections = append(book.Sections, sec)
      }
    } else { // book without sections
      for _, pieceID := range rawBook.Pieces {
        if piece, ok := pieceMap[pieceID]; ok {
          book.Pieces = append(book.Pieces, piece)
        } else {
          err := fmt.Errorf(
            "piece %v from book %v is not in catalog", pieceID, book.ID,
          )
          return nil, err
        }
      }
    }
    books = append(books, book)
  }
  return books, nil
}

func readBooks(
  bookDir, bookFile string, bookIDs []string, all bool, pieceMap PieceMap,
) ([]Book, error) {
  rawBooks, err := readBookFile(bookDir, bookFile)
  if err != nil {
    return nil, err
  }
  if !all { // select books
    rawBooks, err = selectRawBooks(rawBooks, bookIDs)
    if err != nil {
      return nil, err
    }
  }
  books, err := addPiecesToBooks(rawBooks, pieceMap)
  if err != nil {
    return nil, err
  }
  return books, nil
}

func queryPieces(pieces []Piece, queries map[string]string) ([]Piece, error) {
  match, err := makeMatchPiece(queries)
  if err != nil {
    return nil, err
  }
  selPieces := make([]Piece, 0, 200)
  for _, piece := range pieces {
    if match(piece) {
      selPieces = append(selPieces, piece)
    }
  }
  return selPieces, nil
}

func catError(format string, vals ...any) error {
  return fmt.Errorf("catalog: " + format, vals...)
}

func Play(pc PlayCommand) error {
  pieceMap, err := readPieces("catalog", pc.Catalog)
  if err != nil {
    return catError("%v", err)
  }
  pieces := make([]Piece, 0, 1000)
  if pc.Book { // play books
    books, err := readBooks("meta", "books2.yaml", pc.Books, pc.All, pieceMap)
    if err != nil {
      return catError("%v", err)
    }
    for _, book := range books {
      for _, piece := range book.Pieces {
        pieces = append(pieces, piece)
      }
    }
  } else { // play pieces
    if !pc.All { // select pieces
      for _, pieceID := range pc.Pieces {
        if piece, ok := pieceMap[pieceID]; ok {
          pieces = append(pieces, piece)
        } else {
          return catError("piece %v not in catalog", pieceID)
        }
      }
    } else {
      for _, piece := range pieceMap {
        pieces = append(pieces, piece)
      }
    }
  }
  if len(pc.Queries) > 0 {
    pieces, err = queryPieces(pieces, pc.Queries)
    if err != nil {
      return catError("%v", err)
    }
  }
  litter.Dump(pieces)
  return nil
}
